### ISSUE TITLE
Ability to explicitly choose a project type

### ISSUE THREAD

**Issue by vemv** [clojure-lsp/clojure-lsp#1553]
Hi!

Currently, AIUI there's no single, high-level option to choose the project type that will determine how the classpath is computed.

There are two approches being offered:

* Infer the project type out of the presence of bb.edn, deps.edn, project.clj etc
* Choose a `:classpath-cmd`

The problems being:

* Project type inference can favor the thing I didn't want to be favored
  * e.g. there's bb.edn and deps.edn, and I don't want bb to be picked
    * I may not use it (it's a choice that can vary among team members), or CI doesn't have it installed (and we may want to keep it that way)
* `:classpath-cmd` feels low-level
  * makes me repeat a command (`["clojure" "-A:dev:test" "-Spath"]`) that clojure-lsp should be best qualified to know
  * does not reuse `:source-aliases`

### Desired solution/scenario

* I have both deps.edn and bb.edn files in my project
* I can specify `:project-type :deps-edn`
* I can specify `:source-aliases #{,,,}`
* I don't have to specify a `:classpath-cmd`, which in practice ends up being a `clojure` invocation that uses said `:source-aliases`

### Extra context

I'm using clojure-lsp as a CI-oriented linter (for formatting), as I've happily done in other work contexts over the last year.

Cheers - V

**Comment by ericdallo** [comment]
Thanks for the report, I think makes sense to allow exclude some kind of project, project-specs is indeed a low level setting as a last resource when others are not enough, by default clojure-lsp will consider the classpath from all project deps files found, I think we could support a `:exclude-project-file` or something but not couple with keywords like `:deps` or `:bb` as from clojure-lsp POV, it doesn't know what are those project file deps, only know `classpath-cmds` specified by user or the default it uses. Maybe something like `:exclude-project-file #{"bb.edn"}` which clojure-lsp would check if the classpath-cmd contains that string? WDYT?

**Comment by ericdallo** [comment]
yes, clojure-lsp uses [this](https://github.com/clojure-lsp/clojure-lsp/blob/master/lib/src/clojure_lsp/classpath.clj#L169) as default, so if user has any of those files at project root, clojure-lsp will spawn that command for each one and merge, I know that could cause libs used on both with different versions, it's a known issue.

**Comment by ericdallo** [comment]
Thanks for the suggestions :)
I understand your point but not sure how would affect some projects, say there is a project that has bb scripts and clojure code via deps.edn, ATM clojure-lsp works OOTB finding the classpath for both deps.edn and bb.edn allowing to code on both wihout any specific setting, how would that work if we require a setting for those cases?

**Comment by ericdallo** [comment]
Alright, that looks better, but what would be the value of that setting? a set of keywords like `#{:bb :deps}`?
clojure-lsp only know project-files and its classpath-cmds, so maybe something to match the project-file `#{"bb.edn" "deps.edn"}"`?

**Comment by ericdallo** [comment]
Yes, there are some legacy projects that are deps.edn, lein and bb.edn (clj-kondo for example), one may choose to build a classpath from deps and bb for example.

Alright, I think we could support a setting `project-types #{"..." ...}`, although we already have settings to configure classpath and that may become more complex:

- clojure-lsp suggest the usage of `:source-aliases` setting first to calculate the classpath
- if not enough, try using `:source-paths`
- if not enough, use `:project-specs`

And now we would add a new one which should be used instead of `:project-specs`  probably, I wish we could simplify all this complexity

[ref](https://clojure-lsp.io/settings/#classpath-scan)

**Comment by ericdallo** [comment]
> But the use case of being able to tell clojure-lsp "hey, this is my project type and and these are my aliases, please use that and nothing else" should have been there from the beginning IMO.

I agree, the aliases addition made things really easier.

I think it's ok to check the new `project-types #{}` setting when checking for source-aliases, if a project-specs if specified, ignore it
